src/main/java/com/google/collegeplanner/servlets/SavePlanServlet.java

```diff
@@ -0,0 +1,161 @@
+package com.google.collegeplanner.servlets;
+
+import com.google.api.client.extensions.appengine.http.UrlFetchTransport;
+import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken;
+import com.google.api.client.googleapis.auth.oauth2.GoogleIdToken.Payload;
+import com.google.api.client.googleapis.auth.oauth2.GoogleIdTokenVerifier;
+import com.google.api.client.json.gson.GsonFactory;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.collegeplanner.data.Plan;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.json.simple.JSONArray;
+import org.json.simple.JSONObject;
+import org.json.simple.parser.JSONParser;
+import org.json.simple.parser.ParseException;
+
+/** Servlet that saves and returns a users plans.*/
+@WebServlet("/api/planner/save")
+public class SavePlanServlet extends BaseServlet {
+  GoogleIdTokenVerifier verifier;
+  DatastoreService datastore;
+
+  public SavePlanServlet() {
+    this(DatastoreServiceFactory.getDatastoreService(),
+        new GoogleIdTokenVerifier.Builder(UrlFetchTransport.getDefaultInstance(), new GsonFactory())
+            .setAudience(Collections.singletonList(
+                "267429534228-vvsi2uldmpji3rgs1qd3a41rceciaaaq.apps.googleusercontent.com"))
+            .build());
+  }
+
+  public SavePlanServlet(DatastoreService datastore, GoogleIdTokenVerifier verifier) {
+    this.verifier = verifier;
+    this.datastore = datastore;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    GoogleIdToken idToken;
+    try {
+      idToken = verifier.verify(request.getParameter("idToken"));
+    } catch (GeneralSecurityException e) {
+      respondWithError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, response);
+      return;
+    }
+
+    if (idToken == null) {
+      respondWithError("Invalid user.", HttpServletResponse.SC_BAD_REQUEST, response);
+      return;
+    }
+
+    Payload payload = idToken.getPayload();
+    Query query =
+        new Query("Plan")
+            .setFilter(new FilterPredicate("user", FilterOperator.EQUAL, payload.getEmail()))
```

Comment:
I don't think that its necessary for our purposes. Datastore has composite indexing, but it can get a little complex
([https://cloud.google.com/datastore/docs/concepts/indexes](https://cloud.google.com/datastore/docs/concepts/indexes)). I'm not sure I see a real benefit from implementing it for this, though.